File: apps/desktop/src/lib/tauri/device.ts
import type {
	CreateProfileDeviceInput,
	DeviceValidationResult,
	HardwarePreset,
	ProfileDeviceSettings,
	ResolvedDeviceOverview,
	UpdateProfileDeviceSettingsInput,
} from "@/types/device";

import { invokeCommand } from "./client";

function toCreateDeviceInput(input: CreateProfileDeviceInput) {
	return {
		mode: input.mode,
		platform: input.platform,
		hardware_preset_id: input.hardwarePresetId,
		hardware_concurrency: input.hardwareConcurrency,
		device_memory_gb: input.deviceMemoryGb,
		screen_width: input.screenWidth,
		screen_height: input.screenHeight,
		timezone_mode: input.timezoneMode,
		timezone: input.timezone,
		locale_mode: input.localeMode,
		locale: input.locale,
		webrtc_mode: input.webrtcMode,
	};
}

function toUpdateDeviceInput(input: UpdateProfileDeviceSettingsInput) {
	return {
		mode: input.mode,
		platform: input.platform,
		hardware_concurrency: input.hardwareConcurrency,
		device_memory_gb: input.deviceMemoryGb,
		screen_width: input.screenWidth,
		screen_height: input.screenHeight,
		gpu_mode: input.gpuMode,
		hardware_preset_id: input.hardwarePresetId,
		timezone_mode: input.timezoneMode,
		timezone: input.timezone,
		locale_mode: input.localeMode,
		locale: input.locale,
		webrtc_mode: input.webrtcMode,
	};
}

export const deviceApi = {
	get(profileId: string) {
		return invokeCommand<ProfileDeviceSettings>("profile_device_settings_get", {
			profileId,
		});
	},
	update(profileId: string, input: UpdateProfileDeviceSettingsInput) {
		return invokeCommand<ProfileDeviceSettings>(
			"profile_device_settings_update",
			{ profileId, input: toUpdateDeviceInput(input) },
		);
	},
	regenerate(profileId: string) {
		return invokeCommand<ProfileDeviceSettings>(
			"profile_device_settings_regenerate",
			{ profileId },
		);
	},
	validate(profileId: string) {
		return invokeCommand<DeviceValidationResult>(
			"profile_device_settings_validate",
			{ profileId },
		);
	},
	overview(profileId: string) {
		return invokeCommand<ResolvedDeviceOverview>(
			"profile_device_settings_overview",
			{ profileId },
		);
	},
	listPresets() {
		return invokeCommand<HardwarePreset[]>("device_presets_list");
	},
	toCreatePayload(input: CreateProfileDeviceInput) {
		return toCreateDeviceInput(input);
	},
};

File: apps/desktop/src-tauri/src/commands/device.rs
use tauri::State;

use crate::application::services::DeviceSettingsService;
use crate::domain::device::{
    DeviceValidationResult, HardwarePresetDto, ProfileDeviceSettingsDto, ResolvedDeviceOverviewDto,
    UpdateProfileDeviceSettingsInput,
};
use crate::error::AppError;
use crate::state::AppState;

#[tauri::command]
pub async fn profile_device_settings_get(
    state: State<'_, AppState>,
    profile_id: String,
) -> Result<ProfileDeviceSettingsDto, AppError> {
    DeviceSettingsService::get(&state, &profile_id).await
}

#[tauri::command]
pub async fn profile_device_settings_update(
    state: State<'_, AppState>,
    profile_id: String,
    input: UpdateProfileDeviceSettingsInput,
) -> Result<ProfileDeviceSettingsDto, AppError> {
    DeviceSettingsService::update(&state, &profile_id, input).await
}

#[tauri::command]
pub async fn profile_device_settings_regenerate(
    state: State<'_, AppState>,
    profile_id: String,
) -> Result<ProfileDeviceSettingsDto, AppError> {
    DeviceSettingsService::regenerate(&state, &profile_id).await
}

#[tauri::command]
pub async fn profile_device_settings_validate(
    state: State<'_, AppState>,
    profile_id: String,
) -> Result<DeviceValidationResult, AppError> {
    DeviceSettingsService::validate(&state, &profile_id).await
}

#[tauri::command]
pub async fn profile_device_settings_overview(
    state: State<'_, AppState>,
    profile_id: String,
) -> Result<ResolvedDeviceOverviewDto, AppError> {
    DeviceSettingsService::resolve_overview(&state, &profile_id).await
}

#[tauri::command]
pub async fn device_presets_list() -> Result<Vec<HardwarePresetDto>, AppError> {
    Ok(DeviceSettingsService::list_presets())
}
